docs/upgrade/upgrade-v2-to-v3.md

```diff
@@ -0,0 +1,73 @@
+# Upgrade from v2.x to v3.0
+
+1. Upgrade setup
+
+```shell
+composer setup
+```
+
+2. Upgrade schema
+
+Clean up invalid data before running migrations:
+
+```sql
+DELETE FROM Achievements WHERE GameID = 0;
+UPDATE GameData SET ConsoleID = 101 WHERE ConsoleID = 99;
+```
+
+```shell
+php artisan migrate
+```
+
+This may take some time for tables with a lot of data.
+
+> **Note**
+> Running `composer mfs` (alias for `php artisan migrate:refresh --seed`) will remove all columns
+> and data that might have been migrate from then on.
+> The V1 base tables however remain protected and will not be dropped/truncated.
+> It's advised to not use `composer mfs` if you don't want to lose any data.
+ 
+To roll back the migration (e.g. to switch back to a pre-v3 branch):
+
+```shell
+php artisan migrate:rollback
+```
+
+3. Migrate data
+
+Run the following queries and sync commands below to populate the new columns and tables:
+
+> **Note**
+> Depending on the Awarded table size those queries can take several hours.
+> If you had production dumps imported prior you should use the dumps to speed up the process.
+> Note that those only include data up to 2023-02-01 00:00:00. 
+> Any remaining entries have to be synced manually - see sync commands below.
+
+- [01-player-achievements.sql](upgrade-v2-to-v3-01-player-achievements.sql)
+  - Import https://files.retroachievements.org/db/ra-web-player_achievements.gz
+
+- [02-player-games.sql](upgrade-v2-to-v3-02-player-games.sql)
+  - Import https://files.retroachievements.org/db/ra-web-player_games.gz
+
+- [03-leaderboard-entries.sql](upgrade-v2-to-v3-03-leaderboard-entries.sql)
+  - Import https://files.retroachievements.org/db/ra-web-leaderboard_entries.gz
+
+```shell
+php artisan ra:sync:status
+```
+
+If you imported or prepared any data prior you can update the sync_status to save time. 
+
+```sql
+UPDATE sync_status SET reference = '2023-02-01 00:00:00' WHERE kind = 'leaderboard_entries';
+UPDATE sync_status SET reference = '2023-02-01 00:00:00' WHERE kind = 'player_achievements';
+UPDATE sync_status SET reference = '2023-02-01 00:00:00' WHERE kind = 'player_games';
```

Comment:
How do the sync commands differ from the upgrade queries?

The player-achievements.sql only took an hour and a half to run, where the command took 14 hours. Similarly, the player-games.sql took about two hours to run, where the command took 7 hours to hit 100%, and got killed after ~22 hours. And update-leaderboards.sql took about a minute where the command took 17 minutes.

Reply:
Did you run the sync commands after importing the dumps and updating the sync_status tables?

The upgrade queries aren't meant to be run on staging nor production and were added for documentation purposes. As you mentioned, even running those queries on large tables takes far too long for a reasonable migration window, hence why i prepared those dumps for both local and production use; to serve as the initial seed.

Then the sync commands can be used to do the incremental updates.

The plan is to prepare another set of updated dumps that does not end on 2023-02-01 but something more recent (e.g. until end of may) so the incremental syncs don't have to work through a huge amount of rows initially. After it'S all synced it will only have to deal with small amount of rows per minute where it's ok that the commands take longer and aren't as memory efficient. More sync commands will follow to populate many of the other new columns and tables which will work similarly, but won't have to deal with our largest tables. I think those should follow as additional PRs after v3, this branch only prepares schemas for now and just includes the most important denormalizations (player_games) and aggregations (game metrics) so we can use them as soon as possible.

Reply:
> Did you run the sync commands after importing the dumps and updating the sync_status tables?

I followed the steps in the upgrade document (starting at step 2). I already had a previous functional dump loaded in the database.

I did run the upgrade SQL in step 3, but opted to run the sync commands instead of just updating the sync_status because I wanted the most correct data instead of saving time.